Network/AMQP/Internal.hs

```diff
@@ -581,13 +587,17 @@ openChannel c = do
     closed <- newMVar Nothing
     conss <- newMVar M.empty
     listeners <- newMVar []
+    handlers <- newMVar []
 
     --add new channel to connection's channel map
     newChannel <- modifyMVar (connChannels c) $ \mp -> do
         newChannelID <- allocateChannel (connChanAllocator c)
-        let newChannel = Channel c newInQueue outRes (fromIntegral newChannelID) lastConsTag ca closed conss listeners
-        thrID <- forkIO $ CE.finally (channelReceiver newChannel)
-            (closeChannel' newChannel "closed")
+        let newChannel = Channel c newInQueue outRes (fromIntegral newChannelID) lastConsTag ca closed conss listeners handlers
+        thrID <- forkFinally (channelReceiver newChannel) $ \res -> do
```

Comment:
The idea is the following: in the closure we close the channel straight away and in case there was an exception we call all the registered handlers.
I think that for `forkFinally` the same caveats of `forkIO` applies, which means certain async exceptions will be suppressed by default (e.g. `ThreadKilled`)